### PR TITLE
[FLINK-16883] Scan FLINK_CONF directory for available log4j configura…

### PR DESCRIPTION
**Pull request by fapaul**
…tions

This enables the usage of log4j configurations with different
file endings than properties.

## What is the purpose of the change

This change should allow users who bundle their own Flink distribution to change the default logging behavior with the usage of all configurations types allowed by log4j2.
Before only properties files were supported

## Verifying this change

I verified the changes locally and run the `flink-console.sh` script.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by StephanEwen** [comment]
These changes may break setup assumptions from some users. I am not sure.

Maybe one of the users that has built an internal platform, like @mxm , @tweise or @walterddr could comment here, see if they are concerned this would break something?

File: flink-dist/src/main/flink-bin/bin/config.sh [hunk 1]
```diff
@@ -645,3 +645,20 @@ extractExecutionParams() {
...
+    if [[ ${#LOG4J_CONFIG_FILES[@]} -gt 1 ]]; then
+        echo "[ERROR] Found more than one log4j configuration file: [${LOG4J_CONFIG_FILES[@]}]" 1>&2
+        exit 1
+    fi
+
+    LOG4J_CONFIG=${LOG4J_CONFIG_FILES[0]}
+    if [[ "${LOG4J_CONFIG: -1}" = "*" ]]; then
+        echo "[ERROR] Could not find a valid configuration" 1>&2
+        exit 1
+    fi
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/config.sh:661]: I don't understand this block. If we wanted to check whether there were any `log4j-console.*` files, couldn't we check for an empty array in the previous if-block?

What does `if [[ "${LOG4J_CONFIG: -1}" = "*" ]]; then` test?

Comment (mxm) [reply flink-dist/src/main/flink-bin/bin/config.sh:661]: What if the user uses logback, will this fail then?

**Comment by uce** [comment]
I'm also concerned with this. I'm afraid that it adds another layer of "magic" to the scripts.

The original motivation for the ticket was to give higher precedence to user provided Java environment options, e.g. in [flink-console.sh](https://github.com/apache/flink/blob/master/flink-dist/src/main/flink-bin/bin/flink-console.sh#L73)

```bash
java ... `${FLINK_ENV_JAVA_OPTS} "${log_setting[@]}"`
```
vs.
```bash
java ... `"${log_setting[@]}" ${FLINK_ENV_JAVA_OPTS}`
```

This would allow users to manually overwrite the property when they want to use a custom configuration. As a user I actually expect properties that I specify to have higher precedence.

File: flink-dist/src/main/flink-bin/bin/config.sh (hunk 1 above)
Comment (mxm) [reply flink-dist/src/main/flink-bin/bin/config.sh:661]: (Context: We use logback and do not have a log4j file).

Comment (fapaul) [reply flink-dist/src/main/flink-bin/bin/config.sh:661]: > I don't understand this block. If we wanted to check whether there were any `log4j-console.*` files, couldn't we check for an empty array in the previous if-block?
> 
> What does `if [[ "${LOG4J_CONFIG: -1}" = "*" ]]; then` test?

The problem, I experienced, is that the query `($PATH/log4j-console.*)` returns itself if no matches were found. This means the array will never be empty and therefore the additional check.

> What if the user uses logback, will this fail then?

It should not be affected right now since it is set through a different java configuration setting `-Dlogback.configurationFile`.

Comment (mxm) [reply flink-dist/src/main/flink-bin/bin/config.sh:661]: > The problem, I experienced, is that the query `($PATH/log4j-console.*)` returns itself if no matches were found. This means the array will never be empty and therefore the additional check.

Thanks for clarifying, I think it wouldn't hurt to add a comment to explain that bash behavior.

> It should not be affected right now since it is set through a different java configuration setting `-Dlogback.configurationFile`.

I believe it would fail because the changes _require_ a log4j file to be present.

**Comment by zentol** [comment]
I wouldn't encourage users to overwrite the log4j properties. Any user who has setup something like that will run into issue in 1.11 with the switch to log4j2 and renaming of the property.

A user wanting to use an xml file instead of properties shouldn't have to use an entirely separate way of configuring log4j.
I find it a lot more convenient and consistent if users could just drop their log4j configuration into conf and have Flink handle the rest. If that requires a bit of magic, well then so be it.

Note that we can of course still switch the order of the jvm parameters, effectively being an orthogonal issue.

File: flink-dist/src/main/flink-bin/bin/config.sh (hunk 1 above)
Comment (fapaul) [reply flink-dist/src/main/flink-bin/bin/config.sh:661]: Hmm, I am not familiar with the logback setup but I think these changes do not require more presence of a log4j file than before.

The `findLog4jConfiguration()` call is only used in a subshell which will `exit 1` and does not affect the main shell.  Although I agree the log message `[ERROR] Could not find a valid configuration` is not ideal because, in the case of the logback usage, it is probably intended not to have a log4j configuration. 
It might be better to display a warning rather than an error that users are aware no log4j configuration could be found and it falls back to something else.

Comment (mxm) [reply flink-dist/src/main/flink-bin/bin/config.sh:661]: I didn't see that the function call was inside backticks which lets it run in a subshell. Just making sure the log4j configuration remains optional.

Should we return a default value in case of an error? Otherwise these will be the resulting parameters:

```
-Dlog4j.configuration=file: -Dlog4j.configurationFile=file:
```

Not sure if that could cause an error.

**Comment by fapaul** [comment]
Thanks for all the fruitful feedback. I see both proposed solutions as a bit opposing to each other.

First, with the implemented approach, users would name their log4j configuration accordingly and put it into the config directory expecting it to be used.
On the other hand, the solution to give the provided environment options higher precedence adds its own way of setting different log4j configurations.

In summary, having both changes seems like some confusing ambiguity. So we would need to settle with one of them.

Unfortunately, these changes are more user-facing than initially anticipated. Maybe we need to collect more information on how people use different logging configurations.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 2]
```diff
@@ -58,7 +58,9 @@ esac
 
 FLINK_TM_CLASSPATH=`constructFlinkClassPath`
 
-log_setting=("-Dlog4j.configuration=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlog4j.configurationFile=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlogback.configurationFile=file:${FLINK_CONF_DIR}/logback-console.xml")
+log4j_config=`findLog4jConfiguration ${FLINK_CONF_DIR}`
+
+log_setting=("-Dlog4j.configuration=file:${log4j_config}" "-Dlog4j.configurationFile=file:${log4j_config}" "-Dlogback.configurationFile=file:${FLINK_CONF_DIR}/logback-console.xml")
```
Comment (walterddr) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:63]: Reading on the JIRA ticket: seems the concern is user unable to override log settings original done `flink-console.sh` 
Given that, It seems odd to change only log4j but not logback.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 3]
```diff
@@ -58,7 +58,9 @@ esac
 
 FLINK_TM_CLASSPATH=`constructFlinkClassPath`
 
-log_setting=("-Dlog4j.configuration=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlog4j.configurationFile=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlogback.configurationFile=file:${FLINK_CONF_DIR}/logback-console.xml")
+log4j_config=`findLog4jConfiguration ${FLINK_CONF_DIR}`
```
Comment (walterddr) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:61]: I guess the question is why do we still hard-code `${FLINK_CONF_DIR}` ? is there a reason why a user has to put the log files in the same folder where the original log4j.properties is? if the whole idea is to replace it with something else customized, shouldn't the goal be to support user-provided path & files?

Comment (zentol) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:61]: The knowledge of it being in Flink and "dedicated" to Flink could be useful in the future; e.g., we could think about it being modified at runtime via the WebUI to change the log level. You couldn't do this if the log file is at an arbitrary location (for security reasons).

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 2 above)
Comment (zentol) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:63]: The goal was to make the scripts less strict when it comes to the file type (.properties vs .xml). This is more of a concern for log4j since we use a properties file, whereas xml appears to be the "preferred" way with log4j2.
I haven't heard similar problems existing with logback, but yes, ultimately it could be done for both frameworks.

**Comment by zentol** [comment]
Closing since the effort has stalled.

**Comment by Kars-de-Jong** [comment]
> This would allow users to manually overwrite the property when they want to use a custom configuration. As a user I actually expect properties that I specify to have higher precedence.

+1 on this expectation...
